compiler/noirc_frontend/src/tests.rs

```diff
@@ -1444,3 +1451,328 @@ fn specify_method_types_with_turbofish() {
+        errors[0].0,
+        CompilationError::ResolverError(ResolverError::UnsupportedNumericGenericType { .. }),
+    ));
+}
+
+#[test]
+fn numeric_generic_binary_operation_type_mismatch() {
+    let src = r#"
+    fn foo<let N: Field>() -> bool {
+        let mut check: bool = true;
+        check = N;
+        check
+    }   
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert_eq!(errors.len(), 1);
+    assert!(matches!(
+        errors[0].0,
+        CompilationError::TypeError(TypeCheckError::TypeMismatchWithSource { .. }),
+    ));
+}
+
+#[test]
+fn bool_generic_as_loop_bound() {
+    let src = r#"
+    fn read<let N: bool>() {
+        let mut fields = [0; N];
+        for i in 0..N {
+            fields[i] = i + 1;
+        }
+        assert(fields[0] == 1);
+    }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert_eq!(errors.len(), 2);
+
+    assert!(matches!(
+        errors[0].0,
+        CompilationError::ResolverError(ResolverError::UnsupportedNumericGenericType { .. }),
+    ));
+
+    let CompilationError::TypeError(TypeCheckError::TypeMismatch {
+        expected_typ, expr_typ, ..
+    }) = &errors[1].0
+    else {
+        panic!("Got an error other than a type mismatch");
+    };
+
+    assert_eq!(expected_typ, "Field");
+    assert_eq!(expr_typ, "bool");
+}
+
+#[test]
+fn numeric_generic_in_function_signature() {
+    let src = r#"
+    fn foo<let N: u8>(arr: [Field; N]) -> [Field; N] { arr }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert!(errors.is_empty());
+}
+
+#[test]
+fn numeric_generic_as_struct_field_type() {
+    let src = r#"
+    struct Foo<let N: u64> {
+        a: Field,
+        b: N,
+    }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert_eq!(errors.len(), 1);
+    assert!(matches!(
+        errors[0].0,
+        CompilationError::ResolverError(ResolverError::NumericGenericUsedForType { .. }),
+    ));
+}
+
+#[test]
+fn numeric_generic_as_param_type() {
+    let src = r#"
+    fn foo<let I: Field>(x: I) -> I {
+        let _q: I = 5;
+        x
+    }
+    "#;
+    let errors = get_program_errors_elaborator(src);
+    assert_eq!(errors.len(), 3);
+    // Error from the parameter type
+    assert!(matches!(
+        errors[0].0,
+        CompilationError::ResolverError(ResolverError::NumericGenericUsedForType { .. }),
+    ));
+    // Error from the let statement annotated type
+    assert!(matches!(
+        errors[1].0,
+        CompilationError::ResolverError(ResolverError::NumericGenericUsedForType { .. }),
+    ));
+    // Error from the return type
+    assert!(matches!(
+        errors[2].0,
+        CompilationError::ResolverError(ResolverError::NumericGenericUsedForType { .. }),
+    ));
+}
+
+#[test]
+fn numeric_generic_used_in_nested_type_fail() {
+    let src = r#"
+    struct Foo<let N: u64> {
+        a: Field,
+        b: Bar<N>,
+    }
+    struct Bar<N> {
+        inner: N
+    }
```

Comment:
This test looks good. I didn't see a test checking that we get an error when using a type in an array length position or a numeric position of a struct type - can you add these cases?

Reply:
> when using a type in an array length position

We still want to allow this as we have not banned implicit numeric generics. I have the test `implicit_numeric_generics_elaborator` that checks for the implicit generics warning. 

I also added two tests for the simple cases of:
```rust
    struct Foo<N> {
        a: Field,
        b: [Field; N],
    }
```
and 
```rust
    struct Foo<N> {
        a: Field,
        b: Bar<N>,
    }
    struct Bar<let N: u32> {
        inner: [Field; N]
    }
```
I have also linked this issue (https://github.com/noir-lang/noir/issues/5156) to note that the error should be changed once we fully switch to explicit numeric generics.